core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts

```diff
@@ -0,0 +1,419 @@
+    }
+  }
+
+  async getNextQuestionAsync(): Promise<Question> {
+    return new Promise((resolve, reject) => {
+      this._getNextQuestion(resolve, reject);
+    });
+  }
+
+  submitAnswer(
+      answer: InteractionAnswer,
+      interactionRulesService: InteractionRulesService,
+      successCallback: (
+        nextCard: StateCard,
+        refreshInteraction: boolean,
+        feedbackHtml: string,
+        feedbackAudioTranslations: BindableVoiceovers,
+        refresherExplorationId: string,
+        missingPrerequisiteSkillId: string,
+        remainOnCurrentCard: boolean,
+        taggedSkillMisconceptionId: string,
+        wasOldStateInitial: boolean,
+        isFirstHit: boolean,
+        isFinalQuestion: boolean,
+        focusLabel: string
+      ) => void
+  ): void {
+    const oldState: State = this._currentQuestion.getStateData();
+    const classificationResult: AnswerClassificationResult = (
+      this.answerClassificationService.getMatchingClassificationResult(
+        oldState.name as string, oldState.interaction, answer,
+        interactionRulesService));
+    const answerIsCorrect = classificationResult.outcome.labelledAsCorrect;
+
+    let stateCard: StateCard;
+    let refreshInteraction: boolean = false;
+    let feedbackHtml: string = '';
+    let feedbackAudioTranslations: BindableVoiceovers = {};
+    let refresherExplorationId: string = '';
+    let missingPrerequisiteSkillId: string = '';
+    let remainOnCurrentCard: boolean = false;
+    let taggedSkillMisconceptionId: string = '';
+    let wasOldStateInitial: boolean = false;
+    let isFirstHit: boolean = true;
+    let isFinalQuestion: boolean = false;
+    let focusLabel: string;
+
+    this._numberOfAttemptedQuestions += 1;
+
+    if (answerIsCorrect) {
+      this._diagnosticTestCurrentTopicStatusModel.recordCorrectAttempt(
+        this._currentSkillId);
+    } else {
+      this._diagnosticTestCurrentTopicStatusModel.recordIncorrectAttempt(
+        this._currentSkillId);
+    }
+
+    this.getNextQuestionAsync().then((question: Question) => {
+      stateCard = this.createCard(question);
+      focusLabel = this._focusLabel;
+
+      successCallback(
+        stateCard, refreshInteraction, feedbackHtml,
+        feedbackAudioTranslations, refresherExplorationId,
+        missingPrerequisiteSkillId, remainOnCurrentCard,
+        taggedSkillMisconceptionId, wasOldStateInitial, isFirstHit,
+        isFinalQuestion, focusLabel
+      );
+      return answerIsCorrect;
+    }, () => {
+      // Test is finished.
+      const recommendedTopicIds: string[] = this.getRecommendedTopicIds();
+      this.diagnosticTestPlayerStatusService
+        .onDiagnosticTestSessionCompleted.emit(recommendedTopicIds);
+    });
+  }
+
+  getRecommendedTopicIds(): string[] {
+    let recommendedTopicIds: string[] = [];
+    let failedTopicIds: string[] = this.getFailedTopicIds();
+    let rootTopicIds: string[] = this.getRootTopicIds();
```

Comment:
Explain in a comment what a "root topic ID" is.

Reply:
Added